File: Source/Aladdin.CAPI.ANSI.Java/src/aladdin/capi/ansi/kea/KeyPairGenerator.java
package aladdin.capi.ansi.kea;
import aladdin.capi.*; 
import java.math.*; 
import java.io.*;

///////////////////////////////////////////////////////////////////////////
// Алгоритм генерации ключей KEA
///////////////////////////////////////////////////////////////////////////
public class KeyPairGenerator extends aladdin.capi.software.KeyPairGenerator
{
    // параметры ключей
	private final IParameters parameters;

	// конструктор
	public KeyPairGenerator(Factory factory, 
        SecurityObject scope, IRand rand, IParameters parameters)
    { 
        // сохранить переданные параметры
		super(factory, scope, rand); this.parameters = parameters;
	}
    @Override public KeyPair generate(String keyOID) throws IOException 
	{
        // получить фабрику кодирования
        aladdin.capi.KeyFactory keyFactory = factory().getKeyFactory(keyOID); 

		// определить параметры алгоритма
		BigInteger P = parameters.getP(); BigInteger Q = parameters.getQ();
		BigInteger G = parameters.getG(); BigInteger X; 

        // указать генератор случайных чисел
        try (Random random = new Random(rand())) 
        {
            // вычислить предел генерации
            BigInteger max = Q.subtract(BigInteger.valueOf(2)); 

			// сгенерировать случайное число
            int N = Q.bitLength(); do { X = new BigInteger(N, random); }
            
            // проверить условие генерации
            while (X.compareTo(max) >= 0); X = X.add(BigInteger.ONE); 
        }
		// вычислить открытый ключ
		BigInteger Y = G.modPow(X, P);
        
		// создать открытый ключ и личный ключ 
		IPublicKey publicKey = new PublicKey(keyFactory, parameters, Y); 
        
		// создать личный ключ 
		try (IPrivateKey privateKey = new PrivateKey(
            factory(), scope(), keyOID, parameters, X))
        {
            // вернуть созданную пару ключей
            return new KeyPair(publicKey, privateKey, null); 
        }
	}
}

File: Source/Aladdin.CAPI.ANSI/CAPI/ANSI/KEA/KeyPairGenerator.cs
using System; 

namespace Aladdin.CAPI.ANSI.KEA
{
    ///////////////////////////////////////////////////////////////////////////
    // Алгоритм генерации ключей KEA
    ///////////////////////////////////////////////////////////////////////////
    public class KeyPairGenerator :  Software.KeyPairGenerator
    {
        // фабрика алгоритмов и параметры ключей
		private IParameters parameters;

		// конструктор
		public KeyPairGenerator(CAPI.Factory factory, 
            SecurityObject scope, IRand rand, IParameters parameters) 
            
            // сохранить переданные параметры
            : base(factory, scope, rand) { this.parameters = parameters; }

        public override KeyPair Generate(string keyOID) 
	    {
            // получить фабрику кодирования
            CAPI.KeyFactory keyFactory = Factory.GetKeyFactory(keyOID); 

		    // определить параметры алгоритма
		    Math.BigInteger P = parameters.P; Math.BigInteger Q = parameters.Q;
		    Math.BigInteger G = parameters.G; Math.BigInteger X = null; 

	        // вычислить предел генерации
	        Math.BigInteger max = Q.Subtract(Math.BigInteger.ValueOf(2)); int N = Q.BitLength; 

            // указать генератор случайных чисел
            using (Random random = new Random(Rand))
            do { 
			    // сгенерировать случайное число
                X = new Math.BigInteger(N, random); 
            }
            // проверить условие генерации
	        while (X.CompareTo(max) >= 0); X = X.Add(Math.BigInteger.One); 

		    // вычислить открытый ключ
		    Math.BigInteger Y = G.ModPow(X, P);

		    // создать открытый ключ 
		    IPublicKey publicKey = new PublicKey(keyFactory, parameters, Y);

            // создать личный ключ 
            using (IPrivateKey privateKey = new PrivateKey(
                Factory, Scope, keyOID, parameters, X))
            { 
                // вернуть созданную пару ключей
                return new KeyPair(publicKey, privateKey, null); 
            }
	    }
    }
}
